ext4: make ext4_ext_get_blocks always return <= max_blocks

From: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>

ext4_ext_get_blocks() returns number of blocks allocated with buffer
head unmapped for a read from prealloc space.  This is needed so that
delayed allocation doesn't do block reservation for prealloc space since
the blocks are already resevred on disk.  Fix ext4_ext_get_blocks to not
return greater than max_blocks. since some of the code pathes cannot
handle such a return value.

Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
---

 fs/ext4/extents.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)


Index: linux-2.6.25-rc9/fs/ext4/extents.c
===================================================================
--- linux-2.6.25-rc9.orig/fs/ext4/extents.c	2008-04-14 10:24:49.000000000 -0700
+++ linux-2.6.25-rc9/fs/ext4/extents.c	2008-04-14 10:24:51.000000000 -0700
@@ -2570,8 +2570,18 @@ int ext4_ext_get_blocks(handle_t *handle
 			}
 			if (create == EXT4_CREATE_UNINITIALIZED_EXT)
 				goto out;
-			if (!create)
+			if (!create) {
+				/*
+				 * We have blocks reserved already. We
+				 * return allocated blocks so that delalloc
+				 * won't do block reservation for us. But
+				 * the buffer head will be unmapped so that
+				 * a read from the block return 0
+				 */
+				if (allocated > max_blocks)
+					allocated = max_blocks;
 				goto out2;
+			}
 
 			ret = ext4_ext_convert_to_initialized(handle, inode,
 								path, iblock,
